<a href="https://colab.research.google.com/github/rk928/Ranjit-Kumar/blob/main/Copy_of_ENVS_617_Assignment_3_Data_Wrangling_QA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 3: Data Wrangling

In this assignment, you'll be putting your new data wrangling skills to use. You'll be working with data on threatened species from the OECD Environment Statistics database. This dataset on biodiversity shows numbers of known species and threatened species with the aim of indicating the state of mammals, birds, freshwater fish, reptiles, amphibians and vascular plants.

An acronym that may be helpful to know is IUCN: the International Union for Conservation of Nature.

# Part 1. Proportion of endangered mammal species

In part 1, we'll do a series of data manipulations to find out which country has the highest proportion of endangered species among all mammal species. To get started:

1. Import the Pandas package
2. Read the data into a dataframe using `pd.read_csv()`. The csv is avaialable at `https://raw.githubusercontent.com/envirodatascience/ENVS-617-Class-Data/main/OECD_threatened_species.csv`.

In [274]:
import pandas as pd
import numpy as np
df = pd.read_csv('https://raw.githubusercontent.com/envirodatascience/ENVS-617-Class-Data/main/OECD_threatened_species.csv')

In [275]:
# Put your code here to load the pandas package and read in the data frame

## 1.1. Look at the data
First things first: let's see what this data set looks like. Write code below to examine the overall structure of the dataframe. Look at the first rows of the data, the types of columns, the shape of the data. What columns are there (what are the column names)? How many columns and row are there? What do they represent?

In [276]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2923 entries, 0 to 2922
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   IUCN                   2923 non-null   object 
 1   IUCN Category          2923 non-null   object 
 2   SPEC                   2923 non-null   object 
 3   Species                2923 non-null   object 
 4   COU                    2923 non-null   object 
 5   Country                2923 non-null   object 
 6   Unit Code              2923 non-null   object 
 7   Unit                   2923 non-null   object 
 8   Reference Period Code  0 non-null      float64
 9   Reference Period       0 non-null      float64
 10  Value                  2923 non-null   float64
 11  Flag Codes             5 non-null      object 
 12  Flags                  5 non-null      object 
dtypes: float64(3), object(10)
memory usage: 297.0+ KB


## 1.2. How would you describe the structure of this data in words?



information about species, mainly endangered species is contained in total 13 columns, 2923 rows. However, there are some values as null. Data in 3 columns are float and others have string character.

## 1.3. Filtering the data

Since we're interested in data on mammals, filter the dataframe for only data for mammals and store that in a new `mammals_df`.

In [277]:
mammals_df = df[df['Species'] == 'Mammals']
mammals_df.head()
mammals_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 305 entries, 0 to 2922
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   IUCN                   305 non-null    object 
 1   IUCN Category          305 non-null    object 
 2   SPEC                   305 non-null    object 
 3   Species                305 non-null    object 
 4   COU                    305 non-null    object 
 5   Country                305 non-null    object 
 6   Unit Code              305 non-null    object 
 7   Unit                   305 non-null    object 
 8   Reference Period Code  0 non-null      float64
 9   Reference Period       0 non-null      float64
 10  Value                  305 non-null    float64
 11  Flag Codes             0 non-null      object 
 12  Flags                  0 non-null      object 
dtypes: float64(3), object(10)
memory usage: 33.4+ KB


## 1.4. Reshaping the data

Let's reshape the data so that we can compare categories of the status of mammals within a country. Pivot the dataframe so that it has one row per country, with columns for each IUCN category. You will want to have a table that keeps the columns `Country`, `COU`, and `Species` as they are, and then creates new columns for each category in the existing `IUCN` column. Store that in a new `mammals_df_wide`.

Hint: Try reading the `.pivot()` documentation [here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pivot.html).

Now check the shape of the new wide dataframe:

In [278]:
mammals_df_wide = mammals_df.pivot(index=['Country', 'COU', 'Species'], columns='IUCN', values='Value')
mammals_df_wide.head()

IUCN                   CRITICAL  CRITICAL_IND  ENDANGERED  ENDANGERED_IND  \
Country   COU Species                                                       
Australia AUS Mammals      10.0          10.0        38.0            38.0   
Austria   AUT Mammals       4.0           4.0         6.0             6.0   
Belgium   BEL Mammals       4.0           4.0         2.0             2.0   
Brazil    BRA Mammals      12.0           NaN        43.0             NaN   
Canada    CAN Mammals      11.0          11.0        10.0            10.0   

IUCN                   THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
Country   COU Species                                                         
Australia AUS Mammals       107.0           107.0      380.0          353.0   
Austria   AUT Mammals        27.0            27.0      101.0          101.0   
Belgium   BEL Mammals        18.0            18.0       84.0           65.0   
Brazil    BRA Mammals       110.0             NaN        NaN            NaN   
Canada    CAN Mammals        52.0            52.0      221.0          209.0   

IUCN                   VULNERABLE  VULNERABLE_IND  
Country   COU Species                              
Australia AUS Mammals        59.0            59.0  
Austria   AUT Mammals        17.0            17.0  
Belgium   BEL Mammals        12.0            12.0  
Brazil    BRA Mammals        55.0             NaN  
Canada    CAN Mammals        31.0            31.0

## 1.5. Resetting the index

Depending on how you structured your pivot table, you may have noticed that your DataFrame now contains multiple column headers. This is called a _MultiIndex_ in Pandas. Although it sometimes has its uses, to simplify things for now, turn these into their own regular columns by using `.reset_index()`.

Uncomment and run the line of code below. (If your `mammals_df_wide` already only has a single index column, then you can skip this.)

In [279]:
mammals_df_wide = mammals_df_wide.reset_index()
mammals_df_wide.head()


IUCN    Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  \
0     Australia  AUS  Mammals      10.0          10.0        38.0   
1       Austria  AUT  Mammals       4.0           4.0         6.0   
2       Belgium  BEL  Mammals       4.0           4.0         2.0   
3        Brazil  BRA  Mammals      12.0           NaN        43.0   
4        Canada  CAN  Mammals      11.0          11.0        10.0   

IUCN  ENDANGERED_IND  THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
0               38.0       107.0           107.0      380.0          353.0   
1                6.0        27.0            27.0      101.0          101.0   
2                2.0        18.0            18.0       84.0           65.0   
3                NaN       110.0             NaN        NaN            NaN   
4               10.0        52.0            52.0      221.0          209.0   

IUCN  VULNERABLE  VULNERABLE_IND  
0           59.0            59.0  
1           17.0            17.0  
2           12.0            12.0  
3           55.0             NaN  
4           31.0            31.0

## 1.6. Check for duplicates

Lets now make sure we have no duplicated rows of data. Run code to check whether there are duplicates.

In [280]:
mammals_df_wide.duplicated().sum()

0

## 1.7. Calculating the proportion of endangered mammal species
Create a new column that contains the proportion of mammal species in each country that are categorized as endangered. To to this, simply select your dataframe columns and use the `/` operator. Assign this to a new column.


Ex:
```
df['new_column'] = df.column_1 / df.column_2
```


Verify this worked as expected by looking at one row and making sure the calculation is correct.

In [281]:
mammals_df_wide['endangered_proportion'] = mammals_df_wide.ENDANGERED / mammals_df_wide.TOT_KNOWN
mammals_df_wide.head(2)

IUCN    Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  \
0     Australia  AUS  Mammals      10.0          10.0        38.0   
1       Austria  AUT  Mammals       4.0           4.0         6.0   

IUCN  ENDANGERED_IND  THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
0               38.0       107.0           107.0      380.0          353.0   
1                6.0        27.0            27.0      101.0          101.0   

IUCN  VULNERABLE  VULNERABLE_IND  endangered_proportion  
0           59.0            59.0               0.100000  
1           17.0            17.0               0.059406

## 1.8. Sort your data: Which country has the highest proportion?
Now it's time to answer our question: in this data set, which country has the highest proportion of endangered mammal species?


HINT: when you sort the dataframe, you will need use the argument `ascending=False`

Does this answer make sense? Do you have any hypotheses why? (Just a few sentences.)

In [282]:
mammals_df_wide.dropna( )

IUCN          Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  \
0           Australia  AUS  Mammals      10.0          10.0        38.0   
1             Austria  AUT  Mammals       4.0           4.0         6.0   
2             Belgium  BEL  Mammals       4.0           4.0         2.0   
4              Canada  CAN  Mammals      11.0          11.0        10.0   
5               Chile  CHL  Mammals       3.0           3.0        13.0   
6            Colombia  COL  Mammals       5.0           4.0        18.0   
8      Czech Republic  CZE  Mammals      10.0          10.0         1.0   
9             Denmark  DNK  Mammals       1.0           1.0         2.0   
11            Finland  FIN  Mammals       1.0           1.0         4.0   
13            Germany  DEU  Mammals       7.0           7.0        10.0   
14             Greece  GRC  Mammals       3.0           3.0        13.0   
15            Hungary  HUN  Mammals       5.0           5.0         3.0   
21              Korea  KOR  Mammals       1.0           0.0         4.0   
22             Latvia  LVA  Mammals       0.0           0.0         3.0   
24             Mexico  MEX  Mammals       9.0           7.0        50.0   
25        Netherlands  NLD  Mammals       2.0           2.0         3.0   
27             Norway  NOR  Mammals       5.0           5.0         5.0   
28             Poland  POL  Mammals       2.0           2.0        10.0   
29           Portugal  PRT  Mammals      11.0          11.0         7.0   
31    Slovak Republic  SVK  Mammals       2.0           2.0         6.0   
34             Sweden  SWE  Mammals       5.0           5.0         3.0   
35        Switzerland  CHE  Mammals       5.0           5.0        13.0   
37     United Kingdom  GBR  Mammals       2.0           2.0         4.0   

IUCN  ENDANGERED_IND  THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
0               38.0       107.0           107.0      380.0          353.0   
1                6.0        27.0            27.0      101.0          101.0   
2                2.0        18.0            18.0       84.0           65.0   
4               10.0        52.0            52.0      221.0          209.0   
5               13.0        36.0            36.0      185.0          162.0   
6                3.0        55.0            13.0      528.0           58.0   
8                1.0        15.0            15.0       95.0           92.0   
9                2.0         9.0            10.0       71.0           54.0   
11               4.0         7.0             7.0       75.0           67.0   
13              10.0        39.0            39.0      107.0           98.0   
14              13.0        29.0            29.0      115.0          113.0   
15               3.0        14.0            14.0       87.0           76.0   
21               0.0        14.0             0.0      125.0            1.0   
22               3.0         5.0             5.0       62.0           54.0   
24              20.0       150.0            82.0      564.0          162.0   
25               3.0        11.0            11.0       53.0           53.0   
27               5.0        17.0            17.0       89.0           73.0   
28              10.0        13.0            13.0      112.0          100.0   
29               7.0        31.0            31.0      158.0          108.0   
31               6.0        20.0            20.0       90.0           85.0   
34               3.0        14.0            14.0       65.0           65.0   
35              13.0        34.0            34.0       93.0           81.0   
37               4.0        11.0            11.0      107.0           62.0   

IUCN  VULNERABLE  VULNERABLE_IND  endangered_proportion  
0           59.0            59.0               0.100000  
1           17.0            17.0               0.059406  
2           12.0            12.0               0.023810  
4           31.0            31.0               0.045249  
5           20.0            20.0    

In [283]:
mammals_df_wide.sort_values(by='endangered_proportion', ascending=False)

IUCN          Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  \
32           Slovenia  SVN  Mammals       NaN           NaN        19.0   
35        Switzerland  CHE  Mammals       5.0           5.0        13.0   
14             Greece  GRC  Mammals       3.0           3.0        13.0   
0           Australia  AUS  Mammals      10.0          10.0        38.0   
13            Germany  DEU  Mammals       7.0           7.0        10.0   
28             Poland  POL  Mammals       2.0           2.0        10.0   
24             Mexico  MEX  Mammals       9.0           7.0        50.0   
20              Japan  JPN  Mammals      13.0           NaN        12.0   
5               Chile  CHL  Mammals       3.0           3.0        13.0   
31    Slovak Republic  SVK  Mammals       2.0           2.0         6.0   
19              Italy  ITA  Mammals       2.0           NaN         8.0   
1             Austria  AUT  Mammals       4.0           4.0         6.0   
25        Netherlands  NLD  Mammals       2.0           2.0         3.0   
27             Norway  NOR  Mammals       5.0           5.0         5.0   
11            Finland  FIN  Mammals       1.0           1.0         4.0   
22             Latvia  LVA  Mammals       0.0           0.0         3.0   
34             Sweden  SWE  Mammals       5.0           5.0         3.0   
4              Canada  CAN  Mammals      11.0          11.0        10.0   
29           Portugal  PRT  Mammals      11.0          11.0         7.0   
23          Lithuania  LTU  Mammals       NaN           NaN         3.0   
37     United Kingdom  GBR  Mammals       2.0           2.0         4.0   
15            Hungary  HUN  Mammals       5.0           5.0         3.0   
33              Spain  ESP  Mammals       2.0           2.0         6.0   
6            Colombia  COL  Mammals       5.0           4.0        18.0   
21              Korea  KOR  Mammals       1.0           0.0         4.0   
12             France  FRA  Mammals       3.0           NaN         4.0   
7          Costa Rica  CRI  Mammals       1.0           NaN         7.0   
9             Denmark  DNK  Mammals       1.0           1.0         2.0   
2             Belgium  BEL  Mammals       4.0           4.0         2.0   
10            Estonia  EST  Mammals       2.0           2.0         1.0   
8      Czech Republic  CZE  Mammals      10.0          10.0         1.0   
3              Brazil  BRA  Mammals      12.0           NaN        43.0   
16            Iceland  ISL  Mammals       NaN           NaN         NaN   
17            Ireland  IRL  Mammals       NaN           NaN         NaN   
18             Israel  ISR  Mammals       NaN          15.0         NaN   
26        New Zealand  NZL  Mammals       NaN           6.0         NaN   
30             Russia  RUS  Mammals       NaN           NaN         NaN   
36             Turkey  TUR  Mammals       NaN           NaN         NaN   

IUCN  ENDANGERED_IND  THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
32              19.0        34.0            34.0       89.0            NaN   
35              13.0        34.0            34.0       93.0           81.0   
14              13.0        29.0            29.0      115.0          113.0   
0               38.0       107.0           107.0      380.0          353.0   
13              10.0        39.0            39.0      107.0           98.0   
28              10.0        13.0            13.0      112.0          100.0   
24              20.0       150.0            82.0      564.0          162.0   
20               NaN        34.0             NaN      160.0            NaN   
5               13.0        36.0            36.0      185.0          162.0   
31               6.0        20.0            20.0       90.0           85.0   
19               NaN        23.0             NaN      126.0            NaN   
1                6.0        27.0            27.0      101.0          101.0   
25               3.0        11.0            11.0       53.0 

In [284]:
mammals_df_wide.sort_values(by='endangered_proportion', ascending=False).head(1)

IUCN   Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  \
32    Slovenia  SVN  Mammals       NaN           NaN        19.0   

IUCN  ENDANGERED_IND  THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
32              19.0        34.0            34.0       89.0            NaN   

IUCN  VULNERABLE  VULNERABLE_IND  endangered_proportion  
32          15.0            15.0               0.213483

# Part 2. Protected areas
Now let's compare the proportion of endangered mammal species with protected land areas in those countries. First, we'll import a DataFrame with this information. The values in this DataFrame are the percent of total land area that is protected in the country.

In [285]:
url = 'https://raw.githubusercontent.com/envirodatascience/ENVS-617-Class-Data/main/OECD_protected_areas.csv'

In [286]:
df = pd.read_csv(url)
df.head()

Country      Value
0       AUSTRALIA  20.339394
1         AUSTRIA  29.218920
2         BELGIUM  15.559029
3          CANADA  11.904639
4  CZECH REPUBLIC  21.910790

In [287]:
df['Country'] = df['Country'].str.title()

In [288]:
df.rename(columns={'Value': 'Percent_Protected'}, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37 entries, 0 to 36
Data columns (total 2 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Country            37 non-null     object 
 1   Percent_Protected  37 non-null     float64
dtypes: float64(1), object(1)
memory usage: 724.0+ bytes


In [289]:
df.head()

Country  Percent_Protected
0       Australia          20.339394
1         Austria          29.218920
2         Belgium          15.559029
3          Canada          11.904639
4  Czech Republic          21.910790

## 2.1. Look at the data
Again, look at the first few rows of the data and the structure of the data. What data types do the columns have?

In [290]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37 entries, 0 to 36
Data columns (total 2 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Country            37 non-null     object 
 1   Percent_Protected  37 non-null     float64
dtypes: float64(1), object(1)
memory usage: 724.0+ bytes


It has name of the country stored as string and percent of reserved area as float, data type.

## 2.2. Merge with the species data

Merge this DataFrame with `mammals_df_wide` so that for every country in `mammals_df_wide`, we have the protected land area %. This way, we can compare the proportion of endangered mammal species with percent protected areas. You will need to decide whether to use a left join, inner join, or outer join.

Hint: To match rows of two dataframes, the values in the key column need to be identical. How can we transform one or both key colums so that we have identical matches?

In [291]:

new_df = mammals_df_wide.merge(df, on='Country', how='outer')
new_df.head()
new_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Country                40 non-null     object 
 1   COU                    38 non-null     object 
 2   Species                38 non-null     object 
 3   CRITICAL               30 non-null     float64
 4   CRITICAL_IND           27 non-null     float64
 5   ENDANGERED             32 non-null     float64
 6   ENDANGERED_IND         29 non-null     float64
 7   THREATENED             35 non-null     float64
 8   THREATENED_IND         29 non-null     float64
 9   TOT_KNOWN              35 non-null     float64
 10  TOT_KNOWN_IND          28 non-null     float64
 11  VULNERABLE             32 non-null     float64
 12  VULNERABLE_IND         28 non-null     float64
 13  endangered_proportion  31 non-null     float64
 14  Percent_Protected      37 non-null     float64
dtypes: float

## 2.3. Checking the merged DataFrame

After merging the two DataFrames, do some checks to make sure you were able to match all the rows. Get `.info()` on the merged data and then check the first rows.



In [292]:
new_df.info()
new_df.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Country                40 non-null     object 
 1   COU                    38 non-null     object 
 2   Species                38 non-null     object 
 3   CRITICAL               30 non-null     float64
 4   CRITICAL_IND           27 non-null     float64
 5   ENDANGERED             32 non-null     float64
 6   ENDANGERED_IND         29 non-null     float64
 7   THREATENED             35 non-null     float64
 8   THREATENED_IND         29 non-null     float64
 9   TOT_KNOWN              35 non-null     float64
 10  TOT_KNOWN_IND          28 non-null     float64
 11  VULNERABLE             32 non-null     float64
 12  VULNERABLE_IND         28 non-null     float64
 13  endangered_proportion  31 non-null     float64
 14  Percent_Protected      37 non-null     float64
dtypes: float

Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  \
0  Australia  AUS  Mammals      10.0          10.0        38.0   
1    Austria  AUT  Mammals       4.0           4.0         6.0   

   ENDANGERED_IND  THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
0            38.0       107.0           107.0      380.0          353.0   
1             6.0        27.0            27.0      101.0          101.0   

   VULNERABLE  VULNERABLE_IND  endangered_proportion  Percent_Protected  
0        59.0            59.0               0.100000          20.339394  
1        17.0            17.0               0.059406          29.218920

## 2.4 Missing Values


Lets check if any rows have missing data. Display the any rows with missing data.

In [293]:
new_df.isnull().sum()

Country                   0
COU                       2
Species                   2
CRITICAL                 10
CRITICAL_IND             13
ENDANGERED                8
ENDANGERED_IND           11
THREATENED                5
THREATENED_IND           11
TOT_KNOWN                 5
TOT_KNOWN_IND            12
VULNERABLE                8
VULNERABLE_IND           12
endangered_proportion     9
Percent_Protected         3
dtype: int64

In [294]:
new_df[new_df.isnull().any(axis=1)]

Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  \
3          Brazil  BRA  Mammals      12.0           NaN        43.0   
7      Costa Rica  CRI  Mammals       1.0           NaN         7.0   
10        Estonia  EST  Mammals       2.0           2.0         1.0   
12         France  FRA  Mammals       3.0           NaN         4.0   
16        Iceland  ISL  Mammals       NaN           NaN         NaN   
17        Ireland  IRL  Mammals       NaN           NaN         NaN   
18         Israel  ISR  Mammals       NaN          15.0         NaN   
19          Italy  ITA  Mammals       2.0           NaN         8.0   
20          Japan  JPN  Mammals      13.0           NaN        12.0   
23      Lithuania  LTU  Mammals       NaN           NaN         3.0   
24     Luxembourg  NaN      NaN       NaN           NaN         NaN   
27    New Zealand  NZL  Mammals       NaN           6.0         NaN   
31         Russia  RUS  Mammals       NaN           NaN         NaN   
33       Slovenia  SVN  Mammals       NaN           NaN        19.0   
34          Spain  ESP  Mammals       2.0           2.0         6.0   
37         Turkey  TUR  Mammals       NaN           NaN         NaN   
39  United States  NaN      NaN       NaN           NaN         NaN   

    ENDANGERED_IND  THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
3              NaN       110.0             NaN        NaN            NaN   
7              NaN        18.0             NaN      241.0            NaN   
10             1.0         3.0             3.0       66.0           60.0   
12             NaN        17.0             NaN      125.0            NaN   
16             NaN         NaN             NaN        4.0            NaN   
17             NaN         1.0             NaN       57.0           46.0   
18            27.0         NaN            61.0        NaN          105.0   
19             NaN        23.0             NaN      126.0            NaN   
20             NaN        34.0             NaN      160.0            NaN   
23             3.0         7.0             7.0       68.0           59.0   
24             NaN         NaN             NaN        NaN            NaN   
27             2.0         NaN            10.0       65.0           37.0   
31             NaN        34.0             NaN        NaN            NaN   
33            19.0        34.0            34.0       89.0            NaN   
34             6.0        26.0            26.0      174.0            NaN   
37             NaN        23.0             NaN      173.0            NaN   
39             NaN         NaN             NaN        NaN            NaN   

    VULNERABLE  VULNERABLE_IND  endangered_proportion  Percent_Protected  
3         55.0             NaN                    NaN                NaN  
7         10.0             NaN               0.029046          30.630149  
10         NaN             NaN               0.015152          20.043359  
12        10.0             NaN               0.032000          27.254376  
16         NaN             NaN                    NaN          20.459280  
17         1.0             NaN                    NaN          14.217878  
18         NaN            19.0                    NaN          23.934115  
19        13.0             NaN               0.063492          21.487257  
20         9.0             NaN               0.075000          21.372727  
23         4.0             4.0               0.044118          17.004816  
24         NaN             NaN                    NaN          51.091403  
27         NaN             2.0                    NaN          33.363157  
31         NaN             NaN                    NaN                NaN  
33        15.0            15.0               0.213483          40.455267  
34        18.0            18.0               0.034483          28.128468  
37         NaN             NaN                    NaN                NaN  
39         NaN             NaN                    NaN          11.746077

Filter the merged dataframe to look at Brazil, Russia, and Turkey and display the data for those three countries:

In [295]:
new_df[(new_df['Country'] == 'Brazil') | (new_df['Country'] == 'Russia') | (new_df['Country'] == 'Turkey')]

Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  ENDANGERED_IND  \
3   Brazil  BRA  Mammals      12.0           NaN        43.0             NaN   
31  Russia  RUS  Mammals       NaN           NaN         NaN             NaN   
37  Turkey  TUR  Mammals       NaN           NaN         NaN             NaN   

    THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  VULNERABLE  \
3        110.0             NaN        NaN            NaN        55.0   
31        34.0             NaN        NaN            NaN         NaN   
37        23.0             NaN      173.0            NaN         NaN   

    VULNERABLE_IND  endangered_proportion  Percent_Protected  
3              NaN                    NaN                NaN  
31             NaN                    NaN                NaN  
37             NaN                    NaN                NaN

What do you notice for the values for those countries? Do you have any thoughts on why this might be?

most probabaly data regarding endangered species in these countries are not availale with full details or not entered propoerly in the table.

## 2.5 Displaying the Results
To wrap things up, let's display the data we just cleaned in a sorted table.

First, sort your table by the proportion of endangered mammals in each country so that the country with the highest proportion is first (descending order).



In [296]:
new_df.sort_values(by='endangered_proportion', ascending=False)

Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  \
33         Slovenia  SVN  Mammals       NaN           NaN        19.0   
36      Switzerland  CHE  Mammals       5.0           5.0        13.0   
14           Greece  GRC  Mammals       3.0           3.0        13.0   
0         Australia  AUS  Mammals      10.0          10.0        38.0   
13          Germany  DEU  Mammals       7.0           7.0        10.0   
29           Poland  POL  Mammals       2.0           2.0        10.0   
25           Mexico  MEX  Mammals       9.0           7.0        50.0   
20            Japan  JPN  Mammals      13.0           NaN        12.0   
5             Chile  CHL  Mammals       3.0           3.0        13.0   
32  Slovak Republic  SVK  Mammals       2.0           2.0         6.0   
19            Italy  ITA  Mammals       2.0           NaN         8.0   
1           Austria  AUT  Mammals       4.0           4.0         6.0   
26      Netherlands  NLD  Mammals       2.0           2.0         3.0   
28           Norway  NOR  Mammals       5.0           5.0         5.0   
11          Finland  FIN  Mammals       1.0           1.0         4.0   
22           Latvia  LVA  Mammals       0.0           0.0         3.0   
35           Sweden  SWE  Mammals       5.0           5.0         3.0   
4            Canada  CAN  Mammals      11.0          11.0        10.0   
30         Portugal  PRT  Mammals      11.0          11.0         7.0   
23        Lithuania  LTU  Mammals       NaN           NaN         3.0   
38   United Kingdom  GBR  Mammals       2.0           2.0         4.0   
15          Hungary  HUN  Mammals       5.0           5.0         3.0   
34            Spain  ESP  Mammals       2.0           2.0         6.0   
6          Colombia  COL  Mammals       5.0           4.0        18.0   
21            Korea  KOR  Mammals       1.0           0.0         4.0   
12           France  FRA  Mammals       3.0           NaN         4.0   
7        Costa Rica  CRI  Mammals       1.0           NaN         7.0   
9           Denmark  DNK  Mammals       1.0           1.0         2.0   
2           Belgium  BEL  Mammals       4.0           4.0         2.0   
10          Estonia  EST  Mammals       2.0           2.0         1.0   
8    Czech Republic  CZE  Mammals      10.0          10.0         1.0   
3            Brazil  BRA  Mammals      12.0           NaN        43.0   
16          Iceland  ISL  Mammals       NaN           NaN         NaN   
17          Ireland  IRL  Mammals       NaN           NaN         NaN   
18           Israel  ISR  Mammals       NaN          15.0         NaN   
24       Luxembourg  NaN      NaN       NaN           NaN         NaN   
27      New Zealand  NZL  Mammals       NaN           6.0         NaN   
31           Russia  RUS  Mammals       NaN           NaN         NaN   
37           Turkey  TUR  Mammals       NaN           NaN         NaN   
39    United States  NaN      NaN       NaN           NaN         NaN   

    ENDANGERED_IND  THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
33            19.0        34.0            34.0       89.0            NaN   
36            13.0        34.0            34.0       93.0           81.0   
14            13.0        29.0            29.0      115.0          113.0   
0             38.0       107.0           107.0      380.0          353.0   
13            10.0        39.0            39.0      107.0           98.0   
29            10.0        13.0            13.0      112.0          100.0   
25            20.0       150.0            82.0      564.0          162.0   
20             NaN        34.0             NaN      160.0            NaN   
5             13.0        36.0            36.0      185.0          162.0   
32             6.0        20.0            20.0       90.0           85.0   
19             NaN        23.0             NaN      126.0            NaN   
1              6.0        27.0            27.0      101.0          101.0   
26             3.0        11.0

## 2.6 Describe the data

Output the count, mean, standard deviation, min, and percentiles and max for the `proportion_endangered` and `percent_protected`. What is the max value for each?

Make a histogram for each of these columns. Describe what you observe in the two graphs.

In [297]:
print(new_df['endangered_proportion'].describe())

count    31.000000
mean      0.060581
std       0.040995
min       0.010526
25%       0.034287
50%       0.048387
75%       0.072635
max       0.213483
Name: endangered_proportion, dtype: float64


In [300]:
new_df.sort_values(by='endangered_proportion', ascending=False)

Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  \
33         Slovenia  SVN  Mammals       NaN           NaN        19.0   
36      Switzerland  CHE  Mammals       5.0           5.0        13.0   
14           Greece  GRC  Mammals       3.0           3.0        13.0   
0         Australia  AUS  Mammals      10.0          10.0        38.0   
13          Germany  DEU  Mammals       7.0           7.0        10.0   
29           Poland  POL  Mammals       2.0           2.0        10.0   
25           Mexico  MEX  Mammals       9.0           7.0        50.0   
20            Japan  JPN  Mammals      13.0           NaN        12.0   
5             Chile  CHL  Mammals       3.0           3.0        13.0   
32  Slovak Republic  SVK  Mammals       2.0           2.0         6.0   
19            Italy  ITA  Mammals       2.0           NaN         8.0   
1           Austria  AUT  Mammals       4.0           4.0         6.0   
26      Netherlands  NLD  Mammals       2.0           2.0         3.0   
28           Norway  NOR  Mammals       5.0           5.0         5.0   
11          Finland  FIN  Mammals       1.0           1.0         4.0   
22           Latvia  LVA  Mammals       0.0           0.0         3.0   
35           Sweden  SWE  Mammals       5.0           5.0         3.0   
4            Canada  CAN  Mammals      11.0          11.0        10.0   
30         Portugal  PRT  Mammals      11.0          11.0         7.0   
23        Lithuania  LTU  Mammals       NaN           NaN         3.0   
38   United Kingdom  GBR  Mammals       2.0           2.0         4.0   
15          Hungary  HUN  Mammals       5.0           5.0         3.0   
34            Spain  ESP  Mammals       2.0           2.0         6.0   
6          Colombia  COL  Mammals       5.0           4.0        18.0   
21            Korea  KOR  Mammals       1.0           0.0         4.0   
12           France  FRA  Mammals       3.0           NaN         4.0   
7        Costa Rica  CRI  Mammals       1.0           NaN         7.0   
9           Denmark  DNK  Mammals       1.0           1.0         2.0   
2           Belgium  BEL  Mammals       4.0           4.0         2.0   
10          Estonia  EST  Mammals       2.0           2.0         1.0   
8    Czech Republic  CZE  Mammals      10.0          10.0         1.0   
3            Brazil  BRA  Mammals      12.0           NaN        43.0   
16          Iceland  ISL  Mammals       NaN           NaN         NaN   
17          Ireland  IRL  Mammals       NaN           NaN         NaN   
18           Israel  ISR  Mammals       NaN          15.0         NaN   
24       Luxembourg  NaN      NaN       NaN           NaN         NaN   
27      New Zealand  NZL  Mammals       NaN           6.0         NaN   
31           Russia  RUS  Mammals       NaN           NaN         NaN   
37           Turkey  TUR  Mammals       NaN           NaN         NaN   
39    United States  NaN      NaN       NaN           NaN         NaN   

    ENDANGERED_IND  THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
33            19.0        34.0            34.0       89.0            NaN   
36            13.0        34.0            34.0       93.0           81.0   
14            13.0        29.0            29.0      115.0          113.0   
0             38.0       107.0           107.0      380.0          353.0   
13            10.0        39.0            39.0      107.0           98.0   
29            10.0        13.0            13.0      112.0          100.0   
25            20.0       150.0            82.0      564.0          162.0   
20             NaN        34.0             NaN      160.0            NaN   
5             13.0        36.0            36.0      185.0          162.0   
32             6.0        20.0            20.0       90.0           85.0   
19             NaN        23.0             NaN      126.0            NaN   
1              6.0        27.0            27.0      101.0          101.0   
26             3.0        11.0

In [298]:
print(new_df['Percent_Protected'].describe())

count    37.000000
mean     23.460017
std       9.430140
min      11.746077
25%      16.936452
50%      21.372727
75%      28.613044
max      51.091403
Name: Percent_Protected, dtype: float64


Now, sort your table by the proportion of land area protected in each country so that the country with the highest proportion is first (descending order).

In [299]:
new_df.sort_values(by='Percent_Protected', ascending=False)

Country  COU  Species  CRITICAL  CRITICAL_IND  ENDANGERED  \
24       Luxembourg  NaN      NaN       NaN           NaN         NaN   
33         Slovenia  SVN  Mammals       NaN           NaN        19.0   
29           Poland  POL  Mammals       2.0           2.0        10.0   
13          Germany  DEU  Mammals       7.0           7.0        10.0   
32  Slovak Republic  SVK  Mammals       2.0           2.0         6.0   
14           Greece  GRC  Mammals       3.0           3.0        13.0   
27      New Zealand  NZL  Mammals       NaN           6.0         NaN   
7        Costa Rica  CRI  Mammals       1.0           NaN         7.0   
1           Austria  AUT  Mammals       4.0           4.0         6.0   
38   United Kingdom  GBR  Mammals       2.0           2.0         4.0   
34            Spain  ESP  Mammals       2.0           2.0         6.0   
12           France  FRA  Mammals       3.0           NaN         4.0   
18           Israel  ISR  Mammals       NaN          15.0         NaN   
30         Portugal  PRT  Mammals      11.0          11.0         7.0   
15          Hungary  HUN  Mammals       5.0           5.0         3.0   
26      Netherlands  NLD  Mammals       2.0           2.0         3.0   
8    Czech Republic  CZE  Mammals      10.0          10.0         1.0   
19            Italy  ITA  Mammals       2.0           NaN         8.0   
20            Japan  JPN  Mammals      13.0           NaN        12.0   
5             Chile  CHL  Mammals       3.0           3.0        13.0   
16          Iceland  ISL  Mammals       NaN           NaN         NaN   
0         Australia  AUS  Mammals      10.0          10.0        38.0   
10          Estonia  EST  Mammals       2.0           2.0         1.0   
22           Latvia  LVA  Mammals       0.0           0.0         3.0   
28           Norway  NOR  Mammals       5.0           5.0         5.0   
6          Colombia  COL  Mammals       5.0           4.0        18.0   
23        Lithuania  LTU  Mammals       NaN           NaN         3.0   
21            Korea  KOR  Mammals       1.0           0.0         4.0   
9           Denmark  DNK  Mammals       1.0           1.0         2.0   
2           Belgium  BEL  Mammals       4.0           4.0         2.0   
35           Sweden  SWE  Mammals       5.0           5.0         3.0   
25           Mexico  MEX  Mammals       9.0           7.0        50.0   
17          Ireland  IRL  Mammals       NaN           NaN         NaN   
11          Finland  FIN  Mammals       1.0           1.0         4.0   
36      Switzerland  CHE  Mammals       5.0           5.0        13.0   
4            Canada  CAN  Mammals      11.0          11.0        10.0   
39    United States  NaN      NaN       NaN           NaN         NaN   
3            Brazil  BRA  Mammals      12.0           NaN        43.0   
31           Russia  RUS  Mammals       NaN           NaN         NaN   
37           Turkey  TUR  Mammals       NaN           NaN         NaN   

    ENDANGERED_IND  THREATENED  THREATENED_IND  TOT_KNOWN  TOT_KNOWN_IND  \
24             NaN         NaN             NaN        NaN            NaN   
33            19.0        34.0            34.0       89.0            NaN   
29            10.0        13.0            13.0      112.0          100.0   
13            10.0        39.0            39.0      107.0           98.0   
32             6.0        20.0            20.0       90.0           85.0   
14            13.0        29.0            29.0      115.0          113.0   
27             2.0         NaN            10.0       65.0           37.0   
7              NaN        18.0             NaN      241.0            NaN   
1              6.0        27.0            27.0      101.0          101.0   
38             4.0        11.0            11.0      107.0           62.0   
34             6.0        26.0            26.0      174.0            NaN   
12             NaN        17.0             NaN      125.0            NaN   
18            27.0         NaN

Scrolling through each of these two tables, what's something you notice? (Write a few sentences - there is not necessarily a 'right' answer here!)


Austria is having largest percentage of endangered mammal as per data. Luxemberg data is apparently incorrectly entered. Slovenia is having largest area as protected area, as per the data.